addons/coupon/models/coupon.py

```diff
@@ -112,6 +112,6 @@ def _check_coupon_code(self, order_date, partner_id, **kwargs):
             message = {'error': _('This coupon is expired (%s).') % (self.code)}
         elif not self.program_id.active:
             message = {'error': _('The coupon program for %s is in draft or closed state') % (self.code)}
-        elif self.partner_id and self.partner_id.id != partner_id:
+        elif self.partner_id and self.partner_id.id != partner_id and self.order_id.partner_id.id != partner_id:
```

Comment:
`order_id` does not exist without `sale_coupon`, please check my first review.

Reply:
@Williambraecky i tried your first review, but the coupon was created before the partner was created, i can't get the correct partner

Reply:
You may update the partner during `action_confirm` I guess for newly created coupons